src/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Windows.cs

```diff
@@ -330,37 +330,109 @@ public static unsafe SocketError Receive(SafeCloseSocket handle, byte[] buffer,
             return SocketError.Success;
         }
 
-        public static SocketError ReceiveMessageFrom(Socket socket, SafeCloseSocket handle, byte[] buffer, int offset, int size, ref SocketFlags socketFlags, Internals.SocketAddress socketAddress, out Internals.SocketAddress receiveAddress, out IPPacketInformation ipPacketInformation, out int bytesTransferred)
+        public static unsafe IPPacketInformation GetIPPacketInformation(Interop.Winsock.ControlData* controlBuffer)
         {
-            ReceiveMessageOverlappedAsyncResult asyncResult = new ReceiveMessageOverlappedAsyncResult(socket, null, null);
-            asyncResult.SetUnmanagedStructures(buffer, offset, size, socketAddress, socketFlags);
+            IPAddress address = IPAddress.None;
+            if (controlBuffer->length != UIntPtr.Zero)
+            {
+                address = new IPAddress((long)controlBuffer->address);
+            }
```

Comment:
fixed.  But I didn't see any specific other occurrences of this pattern -- the IPv6 code is different and can't use this pattern.  Was there some other code you had in mind?